### PR TITLE
Freeze and dedup strings

### PR DESCRIPTION
**Pull request by jhawthorn**
This commit freezes ❄️ and de-duplicates ♻️ the strings parsed from the JSON db.

This removes about 9600 live objects, about 40% of the live objects from this gem.

**Without gemoji loaded**
$ ruby --disable-gems -Ilib -rjson -e "GC.start;p GC.stat(:heap_live_slots)"
5898

**Before de-duplication**
$ ruby --disable-gems -Ilib -rjson -rgemoji -e "GC.start;p GC.stat(:heap_live_slots)"
29479

**After de-duplication**
$ ruby --disable-gems -Ilib -rjson -rgemoji -e "GC.start;p GC.stat(:heap_live_slots)"
19874

### CODE REVIEW

File: lib/emoji.rb [hunk 1]
```diff
@@ -77,15 +77,23 @@ def parse_data_file
...
+        # Ruby >= 2.5 this will freeze and dedup
+        dedup = lambda { |str| -str }
+      else
+        dedup = lambda { |str| str.freeze }
+      end
+
       append_unicode = lambda do |emoji, raw|
         unless TEXT_GLYPHS.include?(raw) || emoji.unicode_aliases.include?(raw)
-          emoji.add_unicode_alias(raw)
+          emoji.add_unicode_alias(dedup.call(raw))
```
Comment (mislav) [inline lib/emoji.rb]: Each unicode alias is supposed to be unique, i.e. two emoji cannot share the same unicode alias. Does this line really benefit from deduping?

File: lib/emoji.rb [hunk 2]
```diff
@@ -104,12 +112,12 @@ def parse_data_file
...
           end
-          raw_emoji.fetch(:tags).each { |tag| emoji.add_tag(tag) }
+          raw_emoji.fetch(:tags).each { |tag| emoji.add_tag(dedup.call(tag)) }
 
-          emoji.category = raw_emoji[:category]
-          emoji.description = raw_emoji[:description]
-          emoji.unicode_version = raw_emoji[:unicode_version]
-          emoji.ios_version = raw_emoji[:ios_version]
+          emoji.category = dedup.call(raw_emoji[:category])
+          emoji.description = dedup.call(raw_emoji[:description])
```
Comment (mislav) [inline lib/emoji.rb]: Each description is unique. Does this really benefit from deduping?

Comment (jhawthorn) [reply lib/emoji.rb]: This dedup _does_ result in fewer objects (not that much fewer, about 400). It looks like description and aliases can sometimes share the same string.

File: lib/emoji.rb (hunk 1 above)
Comment (jhawthorn) [reply lib/emoji.rb]: Removing this dedup adds about 2000 live objects 🤔 Looking into why...

Comment (jhawthorn) [reply lib/emoji.rb]: Ah, this makes sense. We're using the unicode aliases as keys in a hash, for which Ruby will automatically duplicate and freeze those strings.

```
> hash = {}
> str = "foo"
=> "foo"
> hash[str] = "bar"
=> "bar"
> hash.keys[0]
=> "foo"
> hash.keys[0].frozen?
=> true
> str.frozen?
=> false
> hash.keys[0].object_id
=> 70288572681220
> str.object_id
=> 70288572680560
```

So we were already implicitly deduping these for the hash, but were keeping the unfrozen/undeduped copy around in the unicode_aliases array.

**Pull request merged**